[RISCV][MC] Recognise that fcvt.d.s with frm != 0b000 is valid

asb opened the pull request:
This seems to be an issue common to both GCC and LLVM. There are various RISC-V FCVT instructions where the frm field makes no difference to the output as the result is always exact (e.g. fcvt.d.s, fcvt.s.h, fcvt.d.h). As with GCC, we always generate a form of these fcvt instructions where frm=0b000. However, the ISA manual _doesn't_ state that frm values are invalid, and we should ensure we can accept them. This patch does so by adding the frm field to fcvt.d.s and adding an InstAlias so that if no frm is specified, it defaults to rne (0b000).

This patch just corrects fcvt.d.s in order to allow the approach to be reviewed, before applying it to the other affected instructions.

I haven't added tests to llvm/test/MC/Disassembler/RISCV, because it doesn't seem necessary to test there in addition to our usual round-trip tests in llvm/test/MC/RISCV. But feedback is welcome.

Recently added tests ensure that the default `rne` rounding mode is printed as desired.

asb commented:
For convenience, here is a relevant snippet from the ISA manual. fcvt.d.s will never round, but the frm field isn't listed as being fixed to 0b000.
![fcvt](https://github.com/llvm/llvm-project/assets/1900/9097c5ed-0a59-41c4-adec-affe9e9b95b2)

asb commented:
I've edited the original patch description to clarify this, but wanted to add a comment as well for visibility. I had at one time been concerned about the `rne` rounding mode being printed by us and being rejected by GNU as, but that's not actually an issue provided the InstAlias without the rounding mode is used. This does work for rv32d/rv64d but _doesn't_ seem to be working properly for `dinx` for reasons I haven't yet investigated. We also clearly want test coverage for this as well, given the printing or not of `rne` is important.

Apologies for the confusion on that point. A quick review in terms of the general approach (and to ensure you agree there is a bug to be fixed) would be helpful, but this isn't ready for a detailed review until I add test coverage for the rounding mode printing and resolve whatever the issue with the InstAlias is on dinx.

topperc commented:
Do we end up printing the `rne` with no-aliases now? The check lines don't check the whole line so they wouldn't fail if an extra operand started printing.

Would it be better to have new versions of these for these instruction so we can specific a different default function?

```
def FRMArg : AsmOperandClass {                                                   
  let Name = "FRMArg";                                                           
  let RenderMethod = "addFRMArgOperands";                                        
  let ParserMethod = "parseFRMArg";                                              
  let IsOptional = 1;                                                            
  let DefaultMethod = "defaultFRMArgOp";                                         
}                                                                                
                                                                                 
def frmarg : Operand<XLenVT> {                                                   
  let ParserMatchClass = FRMArg;                                                 
  let PrintMethod = "printFRMArg";                                               
  let DecoderMethod = "decodeFRMArg";                                            
} 
```

asb commented:
> Do we end up printing the `rne` with no-aliases now? The check lines don't check the whole line so they wouldn't fail if an extra operand started printing.

The InstAlias _should_ mean no (and is working as expected for D), but it seems it's not working as expected for dinx. Agree I need to update this with proper test coverage for that.

I wondered about the adding another FRMArg variant with a different defaultFRMArgOps but instead, but it seemed a shame to add new tablegen classes for this. Maybe it's a better solution overall though - I'll have more of a play with it.

asb commented:
> > Do we end up printing the `rne` with no-aliases now? The check lines don't check the whole line so they wouldn't fail if an extra operand started printing.
> 
> The InstAlias _should_ mean no (and is working as expected for D), but it seems it's not working as expected for dinx. Agree I need to update this with proper test coverage for that.
> 
> I wondered about the adding another FRMArg variant with a different defaultFRMArgOps but instead, but it seemed a shame to add new tablegen classes for this. Maybe it's a better solution overall though - I'll have more of a play with it.

1) Commit dc1dc60ea8786d4a2462d157842501d76593f9e7 adds test coverage for the printing of the rounding mode with/without aliases enabled.
2) I've rebased this PR on top of it (although rebasing PRs is discouraged in general, rebasing to account for a newly added test is one of the acceptable reasons in our docs)
3) I've added a commit that instead of using InstAlias, introduces new operand types and the required associated functions. It adds a bit more than I'd like, but it does seem better to not have to define the InstAlias. There are other cases to fix (fcvt.d.w, fcvt.d.wu, fcvt.s.h, fcvt.d.h) so this extra complexity will be somewhat amortized by not needing explicit InstAlias definitions for those cases too.

I think this is ready for full review - thanks!

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVInstPrinter.cpp:
```diff
@@ -158,6 +158,16 @@ void RISCVInstPrinter::printFRMArg(const MCInst *MI, unsigned OpNo,
+                                         raw_ostream &O) {
+  auto FRMArg =
+      static_cast<RISCVFPRndMode::RoundingMode>(MI->getOperand(OpNo).getImm());
+  if (PrintAliases && !NoAliases && FRMArg == RISCVFPRndMode::RoundingMode::RNE)
```
Should we skip printing RNE always so that no-aliases output can still be parsed by older tools?

asb replied:
Good idea, I've made that change. I don't imagine no aliases is commonly used outside of our own backend tests, but no harm in being cautious here.

topperc replied:
I know people who like to use it with objdump to see compressed instructions with their `c.` prefix or for other reasons.

topperc reviewed (approved):
LGTM

asb committed:
[RISCV][MC] Recognise that fcvt.d.s with frm != 0b000 is valid

This seems to be an issue common to both GCC and LLVM. There are various
RISC-V FCVT instructions where the frm field makes no difference to the
output as the result is always exact (e.g. fcvt.d.s, fcvt.s.h,
fcvt.d.h). As with GCC, we always generate a form of these fcvt
instructions where frm=0b000. However, the ISA manual _doesn't_ state
that frm values are invalid, and we should ensure we can accept them.
This patch does so by adding the frm field to fcvt.d.s and adding an
InstAlias so that if no frm is specified, it defaults to rne (0b000).

This patch just corrects fcvt.d.s in order to allow the approach to be
reviewed, before applying it to the other affected instructions.

As suggested during review, we avoid printing the rne rounding mode for
affected instructions even when aliases are disabled, so that the output
can still be consumed by old versions of LLVM or the GNU tools.
